### ISSUE TITLE
High training roullout_actor_probs_pearson_corr in training gpt-oss-20B

### ISSUE THREAD

**Issue by yinzhangyue** [verl-project/verl#3894]
### System Info

When training gpt-oss-20B, there is a very high training/roullout_actor_probs_pearson_corr, about 0.5, far beyond 0.005 suggested in https://verl.readthedocs.io/en/latest/faq/faq.html. Also cannot solve by setting actor_rollout_ref.rollout.engine_kwargs.vllm.disable_cascade_attn=True.

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    data.train_files="$gsm8k_train_path" \
    data.val_files="$gsm8k_test_path" \
    data.train_batch_size=256 \
    data.max_prompt_length=512 \
    data.max_response_length=8192 \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    +data.apply_chat_template_kwargs.reasoning_effort=medium \
    actor_rollout_ref.model.path=${model_dir} \
    actor_rollout_ref.rollout.calculate_log_probs=True \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=256 \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=32 \
    actor_rollout_ref.actor.use_kl_loss=True \
    actor_rollout_ref.actor.kl_loss_coef=0.001 \
    actor_rollout_ref.actor.kl_loss_type=low_var_kl \
    actor_rollout_ref.actor.entropy_coeff=0 \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.fsdp_config.param_offload=False \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=False \
    +actor_rollout_ref.actor.fsdp_config.model_dtype=bfloat16 \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=32 \
    actor_rollout_ref.rollout.tensor_model_parallel_size=2 \
    actor_rollout_ref.rollout.name=sglang \
    actor_rollout_ref.rollout.mode=sync \
    actor_rollout_ref.rollout.engine_kwargs.sglang.attention_backend=triton \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.7 \
    actor_rollout_ref.rollout.n=5 \
    actor_rollout_ref.rollout.load_format=safetensors \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=32 \
    actor_rollout_ref.ref.fsdp_config.param_offload=True \
    algorithm.use_kl_in_reward=False \
    trainer.critic_warmup=0 \
    trainer.logger='["console","wandb"]' \
    trainer.project_name='verl_grpo_example_gsm8k_math' \
    trainer.experiment_name='oai_oss_20b_function_rm' \
    trainer.n_gpus_per_node=8 \
    trainer.nnodes=1 \
    trainer.save_freq=50 \
    trainer.test_freq=10 \
    trainer.total_epochs=15
```

### Expected behavior

training/roullout_actor_probs_pearson_corr < 0.005

**Comment by HJSang** [comment]
Hi, 
I am also debugging this now 
here is our previous exp result: https://github.com/volcengine/verl/issues/3794

**Comment by HJSang** [comment]
My current hypothesis is that there is some unknown model implementation issue for gpt-oss models. We need to try either freeze attention or MOE to see if the training takes any effect or not. It is likely the attention issue. May relate to attention sink.

**Comment by yinzhangyue** [comment]
Thank you for your response, really an important problem in RL Training. Maybe the implementation of SGLang? I am still trying to fix it. Looking forward to more discussions^_^

**Comment by HJSang** [comment]
https://docs.unsloth.ai/models/gpt-oss-how-to-run-and-fine-tune/long-context-gpt-oss-training
Seems related to the implementation in transformers. 
Sglang only does rollout. the loss and probs are calculated by FSDP which is using implementation from transformers.

**Comment by yinzhangyue** [comment]
This issue was caused by verl's default FlashAttention2 not supporting GPT-OSS's attention sink mechanism. Thanks to the improvements in https://github.com/volcengine/verl/pull/3978, we now have two working solutions for training GPT-OSS:

**Option 1: Using eager attention implementation**
```python
python train.py +actor_rollout_ref.model.override_config.attn_implementation=eager
```

**Option 2: Using the community Flash Attention kernel** (recommended for better performance)

First, install the required kernel:
```bash
pip install kernels
```

Then run with:
```python
python train.py +actor_rollout_ref.model.override_config.attn_implementation="kernels-community/vllm-flash-attn3"
```

This second option maintains FlashAttention's speed and GPU memory advantages while providing correct probability distributions.

Both solutions successfully bring the `training/rollout_actor_probs_pearson_corr` metric down to acceptable levels (<0.01), resolving the original issue where it was abnormally high at ~0.5.

**Comment by arde171** [comment]
@yinzhangyue `kernels-community/vllm-flash-attn3` didn't implement the attention-sink backward path. 
<img width="801" height="174" alt="Image" src="https://github.com/user-attachments/assets/465d06f0-6869-49ab-a6de-92c28a75b7cc" />

https://docs.unsloth.ai/new/gpt-oss-reinforcement-learning

**Comment by yinzhangyue** [comment]
@arde171 Thank you! So we can only use Eager implementation? Anyway to accelerate Verl GPT-OSS training?

**Comment by Kirrito-k423** [comment]
Thanks for the detailed report. Based on the current public evidence, I would treat this as a correctness workaround being available, but the high-performance GPT-OSS attention-sink training path is still not fully resolved upstream.

Evidence:
- PR #3978 has been merged and makes `actor_rollout_ref.model.override_config.attn_implementation` effective for FSDP workers, so the actor/reference model can be switched away from the default `flash_attention_2` path.
- The current `docs/advance/attention_implementation.rst` documents `+actor_rollout_ref.model.override_config.attn_implementation=eager` as a supported fallback, and `examples/grpo_trainer/run_gpt_oss_20b_fsdp.sh` saves GPT-OSS-20B with `attn_implementation="eager"`.
- In this thread, `kernels-community/vllm-flash-attn3` was called out as not having the attention-sink backward path, so I would not treat it as an already validated training backward replacement for eager.
- PR #4551, which targets attention-sink support under Ulysses sequence-parallel training, is still open; its description also says FA3 attention sink is still under review/testing.

So the conservative recommendation is: for correctness, use eager for GPT-OSS training, e.g. `+actor_rollout_ref.model.override_config.attn_implementation=eager` (and also set the critic override if the run uses a critic). If the rollout/training mismatch is still large, also evaluate rollout correction/TIS via `algorithm.rollout_correction.rollout_is=token` or `sequence` with a threshold such as `2.0`, because rollout correction is the upstream mechanism for distribution mismatch between rollout and training policies. For acceleration, there is not enough public evidence yet to recommend `kernels-community/vllm-flash-attn3` as a correct backward path; the public upstream work to watch is PR #4551 or a future FA3 attention-sink backward implementation.

中文说明：目前公开证据更支持“已有正确性 workaround，但高性能训练路径尚未完全解决”的判断。PR #3978 已经合并，允许 FSDP 训练通过 `actor_rollout_ref.model.override_config.attn_implementation` 覆盖默认 attention；当前文档和 GPT-OSS 示例都指向 `eager` 作为兼容性 fallback。相反，评论里已经指出 `kernels-community/vllm-flash-attn3` 没有 attention-sink backward path，因此不建议把它当成已验证的训练 backward 加速方案。若仍然观察到 rollout/training 概率分布 mismatch，可以同时评估 rollout correction/TIS，例如 `algorithm.rollout_correction.rollout_is=token` 或 `sequence`，threshold 可先用 `2.0`。

This response was generated by a scheduled Codex GPT5.5 issue processing task.
